File: frontend/src/lib/dm-guard.ts
/**
 * dm-guard.ts — DM送信安全ゲート
 *
 * インシデント対策: テストDMが本物の顧客に送信された問題を防止する
 *
 * 1. DM_TEST_MODE: デフォルトON。テスト時はホワイトリスト以外への送信をブロック
 * 2. campaign必須: campaign_idなしのDMは送信拒否
 * 3. 送信先プレビュー: UI側で呼び出す検証関数を提供
 */

// ─── ホワイトリスト（テストモードで許可するユーザー名） ───
export const DM_TEST_WHITELIST = new Set([
  'pojipojipoji',
  'kantou1234',
  'Nekomeem34',
]);

// ─── テストモード判定 ───
export function isDmTestMode(): boolean {
  // DM_TEST_MODE が明示的に "false" or "off" or "0" のときだけ本番モード
  // 未設定 or それ以外の値 → テストモードON（安全側デフォルト）
  const val =
    (typeof process !== 'undefined' && process.env?.DM_TEST_MODE) ||
    (typeof process !== 'undefined' && process.env?.NEXT_PUBLIC_DM_TEST_MODE) ||
    'true';

  return !['false', 'off', '0'].includes(val.toLowerCase());
}

// ─── エラー型 ───
export class DmGuardError extends Error {
  constructor(
    message: string,
    public readonly code: 'TEST_MODE_BLOCKED' | 'CAMPAIGN_REQUIRED',
    public readonly blockedUsers?: string[],
  ) {
    super(message);
    this.name = 'DmGuardError';
  }
}

// ─── 送信先バリデーション（テストモード時にホワイトリスト外をブロック） ───
export function validateDmTargets(usernames: string[]): {
  allowed: string[];
  blocked: string[];
  isTestMode: boolean;
} {
  const isTestMode = isDmTestMode();

  if (!isTestMode) {
    return { allowed: usernames, blocked: [], isTestMode };
  }

  const allowed: string[] = [];
  const blocked: string[] = [];

  for (const name of usernames) {
    if (DM_TEST_WHITELIST.has(name)) {
      allowed.push(name);
    } else {
      blocked.push(name);
    }
  }

  return { allowed, blocked, isTestMode };
}

// ─── campaign必須バリデーション ───
export function validateCampaignRequired(campaign: string | undefined | null): void {
  if (!campaign || campaign.trim() === '') {
    throw new DmGuardError(
      'campaign_idが指定されていません。全てのDM送信にはcampaign_idが必須です。',
      'CAMPAIGN_REQUIRED',
    );
  }
}

// ─── 統合バリデーション（INSERT前に必ず呼ぶ） ───
export function guardDmSend(
  usernames: string[],
  campaign: string | undefined | null,
): { allowed: string[]; blocked: string[]; isTestMode: boolean } {
  // 1. campaign必須チェック
  validateCampaignRequired(campaign);

  // 2. テストモード＋ホワイトリストチェック
  const result = validateDmTargets(usernames);

  if (result.isTestMode && result.blocked.length > 0) {
    const sample = result.blocked.slice(0, 5).join(', ');
    const suffix = result.blocked.length > 5 ? ` 他${result.blocked.length - 5}名` : '';
    throw new DmGuardError(
      `[DM_TEST_MODE] ホワイトリスト外のユーザーへの送信をブロックしました: ${sample}${suffix}（${result.blocked.length}名）。` +
        `許可ユーザー: ${Array.from(DM_TEST_WHITELIST).join(', ')}`,
      'TEST_MODE_BLOCKED',
      result.blocked,
    );
  }

  return result;
}

// ─── テストモード情報（UI表示用） ───
export function getDmGuardStatus(): {
  isTestMode: boolean;
  whitelist: string[];
} {
  return {
    isTestMode: isDmTestMode(),
    whitelist: Array.from(DM_TEST_WHITELIST),
  };
}

File: backend/services/dm_guard.py
"""
dm_guard.py — DM送信安全ゲート（Python版）

インシデント対策: テストDMが本物の顧客に送信された問題を防止する

1. DM_TEST_MODE: デフォルトON。テスト時はホワイトリスト以外への送信をブロック
2. campaign必須: campaign_idなしのDMは送信拒否
"""

import logging
import os

logger = logging.getLogger(__name__)

# ─── ホワイトリスト（テストモードで許可するユーザー名） ───
DM_TEST_WHITELIST = frozenset([
    "pojipojipoji",
    "kantou1234",
    "Nekomeem34",
])


def is_dm_test_mode() -> bool:
    """テストモード判定。未設定はテストモードON（安全側デフォルト）"""
    val = os.environ.get("DM_TEST_MODE", "true").lower()
    return val not in ("false", "off", "0")


class DmGuardError(Exception):
    """DM安全ゲートエラー"""

    def __init__(self, message: str, code: str, blocked_users: list[str] | None = None):
        super().__init__(message)
        self.code = code
        self.blocked_users = blocked_users or []


def validate_dm_targets(user_names: list[str]) -> tuple[list[str], list[str], bool]:
    """
    テストモード時にホワイトリスト外をブロック。

    Returns:
        (allowed, blocked, is_test_mode)
    """
    test_mode = is_dm_test_mode()

    if not test_mode:
        return user_names, [], False

    allowed = [u for u in user_names if u in DM_TEST_WHITELIST]
    blocked = [u for u in user_names if u not in DM_TEST_WHITELIST]

    return allowed, blocked, True


def validate_campaign_required(campaign: str | None) -> None:
    """campaign必須バリデーション"""
    if not campaign or not campaign.strip():
        raise DmGuardError(
            "campaign_idが指定されていません。全てのDM送信にはcampaign_idが必須です。",
            "CAMPAIGN_REQUIRED",
        )


def guard_dm_send(user_names: list[str], campaign: str | None) -> tuple[list[str], list[str], bool]:
    """
    統合バリデーション。dm_send_log INSERT前に必ず呼ぶ。

    Returns:
        (allowed, blocked, is_test_mode)

    Raises:
        DmGuardError: campaign未指定 or テストモードでホワイトリスト外
    """
    # 1. campaign必須チェック
    validate_campaign_required(campaign)

    # 2. テストモード＋ホワイトリストチェック
    allowed, blocked, test_mode = validate_dm_targets(user_names)

    if test_mode and blocked:
        sample = ", ".join(blocked[:5])
        suffix = f" 他{len(blocked) - 5}名" if len(blocked) > 5 else ""
        raise DmGuardError(
            f"[DM_TEST_MODE] ホワイトリスト外のユーザーへの送信をブロック: {sample}{suffix}（{len(blocked)}名）。"
            f"許可ユーザー: {', '.join(sorted(DM_TEST_WHITELIST))}",
            "TEST_MODE_BLOCKED",
            blocked,
        )

    return allowed, blocked, test_mode
